### DISCUSSION TITLE
Talking with a GSM Modem

### DISCUSSION THREAD

**Discussion by WolleKette** [micropython/micropython#13588] [ESP32]
I'm  trying to communicate in Micropython from an ESP32 controller with a USB-GSM module. Simply send an AT via UART and receive something. I'm getting desperate, it just doesn't work.

If I use a GSM module, which I connect to the controller via RX/TX, then it's super easy

```
from machine import UART
uart = UART(n, rx=..., tx=..., baudrate=...)
uart.write('AT\r\n'
uart.read()
```
But I have to use a GSM module which is connected via USB. I have already tried to connect the modem via a USB-FTDI serial adapter. No reaction.

Then I used the ESP32-S3-DevKitC-1U-N8R8 controller which has a built-in USB-UART port. Supposedly connected to UART 0 and RX=GPIO36, TX=GPIO37. No answer here either.

I know that the question is very general, but I have already tried different modules, so it doesn't make much sense to give any details.  I think it's just a basic comprehension problem.

Please, can anyone give me a hint what could be the reason for this?

**Comment by davefes** [comment]
Have you done a `loop-back` test on UART 0 on the ESP32?

**Comment by karfas** [comment]
I assume the USB on your GSM module expects to talk to an USB host and not to another USB device.
USB host mode doesn't exist in micropython yet, but CircuitPython has something - see [their documentation](https://docs.circuitpython.org/en/latest/shared-bindings/usb_host/index.html).

**Comment by kjm1102** [comment]
In the bad old days I could not get a Sierra 320 USB modem to work on a rpi model A. Turns out I was trying to run before I could crawl. Ended up plugging the 320U into a PC with some manufacturers software  as the driver to get familiar with the AT cmd structure. Once I has a feel for the modems foibles getting it to work on the pi went much smoother.

**Comment by WolleKette** [comment]
@karfas: That with the USB host sounds plausible. I will install Circuit Python on an ESP32-S3-DevKit board in the next few days and experiment with it. I will report back...

@kjm1102: What are you trying to tell me?
I connected one of the modems to a Windows PC and was able to communicate via Kitty (Putty) using AT commands. I did it to test the module. You write that you got it to run on a pi. How?

I read somewhere that the rp2 port of Micropython implements TinyUSB, which supports USB Host. But in the MP documentation on GitHub I didn't find anything about it.

**Comment by robert-hh** [comment]
> I read somewhere that the rp2 port of Micropython implements TinyUSB, which supports USB Host. But in the MP documentation on GitHub I didn't find anything about it.

TinyUSB supports host mode, but that is not used by MicroPyhton. There is a ongoing work by @projectgus to support Pyhton coded USB drivers, which would then allow host mode. #9497

**Reply by projectgus** [reply]
So far this work is just for Python on the USB device implementation. The host side is something we'd also like to do, but no work has been done on that side yet.
